twistes/client.py

```diff
@@ -617,8 +621,17 @@ def _perform_request(self, method, path, body=None, params=None):
             raise ElasticsearchException(msg_fmt.format(code=response.code,
                                                         msg=str(content)))
 
-        except (ResponseNeverReceived,
-                CancelledError,
+        except ResponseNeverReceived as e:
+
+            if self._retry_on_timeout and self._max_retries <= 3:
```

Comment:
Don't mutate the state of the client for all requests
pass num_of_retries as a parameter with default value of the self._max_retries 
and call _perform_request with num_of_retries - 1.